File: leetcode/algorithms/tree/binary-tree-level-order-traversal-ii.java
/*
    leetcode
    category: "Algorithms"
    problem title: "Binary Tree Level Order Traversal II"
*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int getTreeHeight(TreeNode root){
        if(root==null){
            return 0;
        }
        
        return 1+Math.max(getTreeHeight(root.left),getTreeHeight(root.right));
    }
    
    private void populateListForLevel(TreeNode node, int level, ArrayList<List<Integer>> levelsList){
        if(node==null){
            return;
        }
        
        populateListForLevel(node.left,level-1,levelsList);
        populateListForLevel(node.right,level-1,levelsList);
        
        levelsList.get(level).add(node.val);
        
    }
    
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        int height = getTreeHeight(root);
        
        ArrayList<List<Integer>> levelsList = new ArrayList<List<Integer>>();
        
        for(int i=0;i<height;++i){
            levelsList.add(new ArrayList<Integer>());
        }
        
        populateListForLevel(root,height-1,levelsList);
        
        return levelsList;
        
    }
}

File: leetcode/algorithms/tree/binary-tree-level-order-traversal-ii.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */



function levelOrder(root: TreeNode | null): number[][] {
    
    const queue: Array<{ 
        node: TreeNode;
        level: number;
    }> = [];
    
    if(root!==null){
        queue.push({
            node: root,
            level: 0,
        })
    }
    
    const result: number[][] = [];
    
    while(queue.length>0){
        const curr = queue.shift();
        
        if(curr.node.left){
            queue.push({
                node: curr.node.left,
                level: curr.level+1,
            })
        }
        
        if(curr.node.right){
            queue.push({
                node: curr.node.right,
                level: curr.level+1,
            })
        }
        
        if(curr.level===result.length){
            result.push([curr.node.val]);
        } else {
            result[curr.level].push(curr.node.val);
        }
        
    }
    
    return result;
          
};